datafusion/src/physical_plan/hash_join.rs

```diff
@@ -1410,6 +1413,45 @@ mod tests {
         assert_batches_sorted_eq!(expected, &batches);
     }
 
+    #[tokio::test]
+    async fn join_outer_multi_batch() {
+        let left = build_table(
+            ("a1", &vec![1, 2, 3]),
+            ("b1", &vec![4, 5, 7]), // 7 does not exist on the right
+            ("c1", &vec![7, 8, 9]),
+        );
+        let right = build_table_two_batches(
+            ("a2", &vec![10, 20, 30]),
+            ("b1", &vec![4, 5, 6]),
+            ("c2", &vec![70, 80, 90]),
+        );
+        let on = &[("b1", "b1")];
+
+        let join = join(left, right, on, &JoinType::Full).unwrap();
+
+        let columns = columns(&join.schema());
+        assert_eq!(columns, vec!["a1", "b1", "c1", "a2", "c2"]);
+
+        let stream = join.execute(0).await.unwrap();
+        let batches = common::collect(stream).await.unwrap();
+
+        let expected = vec![
```

Comment:
This doesn't look like the correct answer to me. FULL OUTER JOIN basically keeps tuples from boths sides that don't match the join criteria.

So we should have 2 rows for `b1=4' and `b1=5` and then 1 row for tuple on the left that has no match with the right, `b1=7` and 1 row for the tuple on the right that has no match with the left `b1=6`

And indeed  postgres returns 4 rows as well


```
alamb=# select * from tleft;
 a1 | b1 | c1 
----+----+----
  1 |  4 |  7
  2 |  5 |  8
  3 |  7 |  9
(3 rows)

alamb=# select * from tright;
 a2 | b2 | c2 
----+----+----
 10 |  4 | 70
  2 |  5 | 80
  3 |  6 | 90
(3 rows)

alamb=# select * from tleft FULL JOIN tright ON (tleft.b1 = tright.b2);
 a1 | b1 | c1 | a2 | b2 | c2 
----+----+----+----+----+----
  1 |  4 |  7 | 10 |  4 | 70
  2 |  5 |  8 |  2 |  5 | 80
    |    |    |  3 |  6 | 90
  3 |  7 |  9 |    |    |   
(4 rows)
```

Reply:
The test here tests having two (equivalent) batches at the right side.
So they will produce two rows for the right rows and one for the left row that doesnt match the right row.

Reply:
To me the results look correct given two equivalent right batches. I included `b2` as column here as well to make it a bit more obvious:

```
            "+----+----+----+----+----+----+",
            "| a1 | b1 | c1 | a2 | b2 | c2 |",
            "+----+----+----+----+----+----+",
            "|    |    |    | 30 | 6  | 90 |",
            "|    |    |    | 30 | 6  | 90 |",
            "| 1  | 4  | 7  | 10 | 4  | 70 |",
            "| 1  | 4  | 7  | 10 | 4  | 70 |",
            "| 2  | 5  | 8  | 20 | 5  | 80 |",
            "| 2  | 5  | 8  | 20 | 5  | 80 |",
            "| 3  | 7  | 9  |    |    |    |",
            "+----+----+----+----+----+----+",
```